Improve injection detection and some refactoring

0x534a opened the pull request:
I suggest to improve the detection of code injection techniques with this PR. It fixes issue #88 partly by contributing new rules for certain injection techniques and by refactoring existing ones.

The suggested changes in this PR are mainly based on https://www.elastic.co/de/blog/ten-process-injection-techniques-technical-survey-common-and-trending-process and include:

- New rule for DLL injection (ATT&CK ID T1055.001)
- New rule for PE injection (ATT&CK ID T1055.002)
- New rules for MBC micro-behavior (thread suspend and resume)
- Refactoring of existing injection detection rules

The test-files repository will be updated accordingly with the added example files of the suggested rules afterwards.

0x534a commented:
Corresponding PR to add test files in the capa-testfiles repo is [#56](https://github.com/fireeye/capa-testfiles/pull/56)

mr-tz commented on host-interaction/thread/create/create-thread.yml:
```diff
@@ -19,6 +19,12 @@ rule:
+        - api: ntdll.NtCreateThreadEx
+        - api: ntdll.ZwCreateThread
+        - api: ntdll.ZwCreateThreadEx
         - and:
           - api: kernel32.GetCurrentProcess
           - api: kernel32.CreateRemoteThread
```
We capture all thread creations here now, so can remove this block?

0x534a replied:
I just left it since I thought it has a deeper sense for a certain rule. It is removed in aa94a33.

mr-tz commented on host-interaction/process/inject/use-thread-execution-hijacking.yml:
```diff
@@ -0,0 +1,25 @@
+rule:
+  meta:
+    name: use thread execution hijacking
```
propose: `hijack thread execution`

0x534a replied:
Fixed in aa94a33

mr-tz commented on host-interaction/process/inject/use-thread-execution-hijacking.yml:
```diff
@@ -0,0 +1,25 @@
+    - and:
+      - optional:
+        - or:
+          - api: kernel32.OpenThread
+          - api: ntdll.NtOpenThread
+          - api: ntdll.ZwOpenThread
```
looks like another good (library) rule candidate `open thread`

0x534a replied:
Fixed in aa94a33

mr-tz commented on host-interaction/process/inject/use-process-replacement.yml:
```diff
@@ -8,11 +8,11 @@ rule:
-          - api: kernel32.ResumeThread
-          - api: ntdll.NtResumeThread
+      - match: write process memory
+      - match: resume thread
```
awesome!

mr-tz commented on host-interaction/process/inject/use-pe-injection.yml:
```diff
@@ -0,0 +1,29 @@
+rule:
+  meta:
+    name: use pe injection
```
propose: `inject PE`

williballenthin replied:
"this function may...."
"inject *a* PE"

0x534a replied:
Fixed in aa94a33

mr-tz commented on host-interaction/process/inject/use-pe-injection.yml:
```diff
@@ -0,0 +1,29 @@
+      - characteristic: loop
+      - optional:
+        - or:
+          - api: kernel32.OpenProcess
+          - api: ntdll.NtOpenProcess
+          - api: ntdll.ZwOpenProcess
```
new (library) rule candidate: `open process`

0x534a replied:
Fixed in aa94a33

mr-tz commented on host-interaction/process/inject/use-pe-injection.yml:
```diff
@@ -0,0 +1,29 @@
+          - api: ntdll.ZwOpenProcess
+          - match: create process
+      - match: allocate RWX memory
+      - basic block:
+        - description: virtual address offset calculation
+        - and:
+          - mnemonic: and
+          - number: 0x0FFF
```
I'm not sure if this is a key part in every implementation of `pe injection`

0x534a replied:
I am also unsure regarding this block of the rule. I simply followed https://www.elastic.co/de/blog/ten-process-injection-techniques-technical-survey-common-and-trending-process in this case in order to write the rule. 

I think, you are the more proficient reverse engineers than me and maybe have more experience regarding pe injection. Do have certain samples in mind that implement pe injection in another way?

I will also do some research on my own but this will take some time.

mr-tz replied:
I say we remove this part.

0x534a replied:
I think, we could then remove the whole rule since there is no real difference to `inject-thread.yml`.

mr-tz replied:
Ah, I see. Let's keep the rule as is then.

mr-tz commented on host-interaction/process/inject/use-dll-injection.yml:
```diff
@@ -0,0 +1,34 @@
+rule:
+  meta:
+    name: use dll injection
```
propose: `inject DLL`

0x534a replied:
Fixed in aa94a33. Also added _a_ as suggested by @williballenthin for the PE injection rule.

mr-tz commented on host-interaction/process/inject/allocate-rwx-memory.yml:
```diff
@@ -21,6 +21,8 @@ rule:
         - api: kernel32.VirtualProtect
```
should we remove the `att&ck` mapping as this general functionality is not necessarily process injection?

0x534a replied:
Fixed in aa94a33

mr-tz commented on host-interaction/process/inject/allocate-rw-memory.yml:
```diff
@@ -0,0 +1,26 @@
+    namespace: host-interaction/process/inject
+    author: 0x534a@mailbox.org
+    scope: basic block
+    att&ck:
+      - Defense Evasion::Process Injection [T1055]
```
Propose to remove this as not specific enough to process injection?

0x534a replied:
Fixed in aa94a33

mr-tz commented on host-interaction/process/inject/allocate-rw-memory.yml:
```diff
@@ -0,0 +1,26 @@
+  features:
+    - and:
+      - or:
+        - api: kernel32.VirtualAlloc
+        - api: kernel32.VirtualAllocEx
+        - api: kernel32.VirtualAllocExNuma
+        - api: kernel32.VirtualProtect
+        - api: kernel32.VirtualProtectEx
+        - api: ntdll.NtAllocateVirtualMemory
+        - api: ntdll.ZwAllocateVirtualMemory
+        - api: ntdll.NtMapViewOfSection
+        - api: ntdll.ZwMapViewOfSection
```
could move these to a library rule `allocate memory` and use a match in the RW and RWX rules

0x534a replied:
Fixed in aa94a33 but there is a linter error. See comment below.

mr-tz reviewed (changes requested):
Nice additions, thank you! Added a couple of improvement ideas and comments.
I prefer that rule names start with a verb other than `use` although we set the example with `process replacement` ourselves ;)

If possible, please rebase to the current master, so the linter action can run successfully using the newly merged test files.

0x534a commented:
Thanks for all your comments. I addressed most of them in the newly provided commits. 

The linter currently fails with the following error:
```
allocate memory 


  FAIL: lib rule not found in lib directory: Move the rule to the `lib` subdirectory of the rules path
```

The `allocate-memory.yml` is already placed in the `lib` directory. Can you help me with fixing this linter error?

0x534a commented:
I noticed that the usage of indefinite articles in rule names is quite inconsistent in the current rule base. E.g. `create-process.yml` is not using an article but `create-a-process-with-modified-io-handles-and-window.yml` is using one. This can be addressed in the future to make the rule base more consistent and to not confuse rule contributors. ;D

mr-tz commented:
> 
> 
> Thanks for all your comments. I addressed most of them in the newly provided commits.
> 
> The linter currently fails with the following error:
> 
> ```
> allocate memory 
> 
> 
>   FAIL: lib rule not found in lib directory: Move the rule to the `lib` subdirectory of the rules path
> ```
> 
> The `allocate-memory.yml` is already placed in the `lib` directory. Can you help me with fixing this linter error?

running this locally, I see the following:
```
allocate RW memory
  FAIL: doesn't match on referenced example: Fix the rule logic or provide a different example

allocate user process RWX memory
  FAIL: doesn't match on referenced example: Fix the rule logic or provide a different example
```

I'll investigate the missing matches and look into fixing the check linter output.

mr-tz commented:
Reran this after the capa fix and getting
```
hijack thread execution 
  FAIL: doesn't match on referenced example: Fix the rule logic or provide a different example
```

The offset doesn't match the sample. Could you take another look, please?

0x534a commented:
Thanks for the update!

Unfortunately, the sample referenced by the rule was not suitable since the memory allocation happens in a separate function. Therefore, I added another sample in the test files repository which matches the rule. Sorry for the extra work.

I also reworked the usage of articles by removing them from the rules and resolved the merge conflict.

0x534a commented:
There was another problem with the example files of the rule `inject-thread.yml`. The example file `Practical Malware Analysis Lab 12-01.exe_:0x4010D0` had no detection since the rule only matched on the allocation of RWX memory but this sample only allocates RW memory. This is fixed in commit cf84ebe.

Thanks for your comments!

mr-tz reviewed (approved):
Awesome, thanks a lot @0x534a! Great additions and improvements.
Happy to see that the linter helps us even more now to keep the rules quality high.

0x534a commented:
That's great news! :)

Thanks for merging my PR and for all your effort in fixing the linter and reviewing my stuff, @mr-tz!
